File: app/vocatio/lib/models/aula.dart
import 'package:vocattio/models/presenca.dart';
import 'package:vocattio/utils/date_formater.dart';

class Aula {
  final String objectId;
  final String turmaId;
  final String codigo;
  final bool aberta;
  final double latitude;
  final double longitude;
  final DateTime dataAbertura;
  final List<Presenca> presentes;
  final DateTime? dataFechamento;

  const Aula({
    required this.objectId,
    required this.turmaId,
    required this.codigo,
    required this.aberta,
    required this.latitude,
    required this.longitude,
    required this.dataAbertura,
    required this.presentes,
    this.dataFechamento,
  });

  factory Aula.fromJson(Map<String, dynamic> json) {
    return Aula(
      objectId: json['_id'] as String,
      turmaId: json['turmaId'] as String,
      codigo: json['codigo'] as String,
      aberta: json['aberta'] as bool,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      dataAbertura: DateParser.parseCustomDate(json['dataAbertura']),
      presentes: (json['presentes'] as List)
        .map((item) => Presenca.fromJson(item as Map<String, dynamic>))
        .toList(),
      dataFechamento: json['dataFechamento'] == null
          ? null
          : DateParser.parseCustomDate(json['dataFechamento']),
    );
  }

}

File: server/src/domain/Aula.java
package src.domain;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class Aula {
    private String _id;
    private String turmaId;
    private String codigo;
    private boolean aberta;
    private Double latitude;
    private Double longitude;
    private Date dataAbertura;
    private List<Presenca> presentes;
    private Date dataFechamento;

    public Aula(){}
    public Aula(String _id, String turmaId, String codigo, boolean aberta, Double latitude,
                Double longitude, Date dataAbertura, List<Presenca> presentes, Date dataFechamento) {
        this._id = _id;
        this.turmaId = turmaId;
        this.codigo = codigo;
        this.aberta = aberta;
        this.latitude = latitude;
        this.longitude = longitude;
        this.dataAbertura = dataAbertura;
        this.presentes = presentes;
        this.dataFechamento = dataFechamento;
    }

    public Date getDataAbertura(){ return this.dataAbertura; }
    public List<Presenca> getPresentes (){return this.presentes;}

    @Override
    public String toString(){
        return "_id: " + this._id +
                "\n turmaId: " + this.turmaId +
                "\n codigo: " + this.codigo +
                "\n aberta: " + this.aberta +
                "\n latitude: " + this.latitude +
                "\n longitude: " + this.longitude +
                "\n dataAbertura: " + this.dataAbertura +
                "\n presentes: " + this.presentes +
                "\n dataFechamento: " + this.dataFechamento;
    }

    @Override
    public boolean equals(Object obj){
        if(obj == this) return true;
        if(obj == null) return false;
        if(this.getClass() != obj.getClass()) return false;
        Aula a = (Aula) obj;
        return this._id != null && this._id.equals(a._id);
    }

    @Override
    public int hashCode(){
        int ret = 1;
        ret = 31 * ret + this._id.hashCode();
        ret = 31 * ret + this.turmaId.hashCode();
        ret = 31 * ret + this.codigo.hashCode();
        ret = 31 * ret + this.latitude.hashCode();
        ret = 31 * ret + this.longitude.hashCode();
        ret = 31 * ret + this.dataAbertura.hashCode();
        ret = 31 * ret + this.presentes.hashCode();
        ret = 31 * ret + this.dataFechamento.hashCode();
        ret = 31 * ret + this.presentes.hashCode();
        return ret;
    }

}
